### PR TITLE
fix(tr6): close three redundant-type-conversion soundness gaps

### LINKED ISSUE #220: redundant-type-conversion (TR6): aggressive-mode hover_passes_gate always passes non-matching, non-union types
## Summary

In `aggressive` mode, `hover_passes_gate()` (`src/pre_commit_hooks/ast_checks/redundant_type_conversion/confidence.py:83-93`) is supposed to widen the exact-match gate to also allow "looser" matches (e.g. `list[str]` where only `Iterable[str]` is required). Instead, for any hover that resolves to a single (non-union) type and isn't an exact match, it unconditionally returns `True` — with no relationship check to the constructor at all:

```python
def hover_passes_gate(hover_text: str | None, level: ConfidenceLevel, constructor: str) -> bool:
    if not hover_text or _is_unreliable(hover_text):
        return False
    if _exact_match(hover_text, constructor):
        return True
    if level is not ConfidenceLevel.AGGRESSIVE:
        return False
    members = _split_top_level_union(hover_text)
    if len(members) > 1:
        return all(_exact_match(member, constructor) for member in members)
    return True  # <-- reached for ANY single-member, non-exact-match hover
```

This means every `constructor(arg)` candidate in aggressive mode reaches the expensive ty diagnostics-diff stage regardless of whether `arg`'s type has any plausible relationship to `constructor` — including cases like wrapping a completely unrelated custom class in `str(...)`.

## Repro

```py
import random
from http import HTTPStatus
from typing import Final

EXTERNAL_SERVICE_ERRORS_STATUS_CODES: Final = frozenset(
    {HTTPStatus.TOO_MANY_REQUESTS, HTTPStatus.INTERNAL_SERVER_ERROR}
)
extra_ignore_response_status_codes: set[HTTPStatus] = set()

def response_status_code() -> int:
    response_status_codes = set(EXTERNAL_SERVICE_ERRORS_STATUS_CODES)
    response_status_codes.update(extra_ignore_response_status_codes)  # type: ignore[arg-type]  # ty: ignore[invalid-argument-type]
    return random.choice(tuple(response_status_codes))
```

```
$ ruff-extra-rules --select=redundant-type-conversion --redundant-type-conversion-level=aggressive repro.py
repro.py:9:29: TR6: Redundant `set(...)` conversion: `ty` sees no difference with or without this wrap here, though the argument's own type is `frozenset[HTTPStatus] (Final)`, not `set` -- verify before removing. ...
```

`frozenset[HTTPStatus]` is not remotely a "looser" match for `set` in the sense the aggressive-mode docstring describes (e.g. `list[str]` vs `Iterable[str]`) — it's a different, incompatible type (no `.update()`). Same shape of false positive for `str(some_custom_exception)` used in a `not in` comparison: `ExtendedClientResponseError` has no relationship to `str` at all, yet passes this gate.

## Why this matters even though the diagnostics-diff is the real safety net

In a correctly-scoped project, the subsequent before/after diagnostics-diff in `decide_candidates()` (`analysis.py`) does reliably catch the `frozenset`/`.update()` mismatch — I confirmed this directly against a freshly-scoped `ty` session (removing `set(...)` introduces a new `unresolved-attribute` diagnostic that isn't present in the baseline). So this gate isn't the only thing standing between "any type" and a report. But it's still doing none of the filtering its own aggressive-mode help text claims ("a looser match, e.g. an already-`list[str]` value passed somewhere only an `Iterable[str]` is required"), and it needlessly sends unrelated-type candidates through an expensive two-pass `ty` round trip. Separately, see #221 for a case where the diagnostics-diff itself misses the difference when `ty`'s analysis of the file is degraded — the two problems compound.

## Suggested fix

`hover_passes_gate`'s aggressive branch should still require *some* structural relationship between the hover type and the constructor (e.g. both sides sharing a common iterable/element relationship, or the hover being a subtype/supertype in the same protocol family) rather than falling through to `True` for any non-union hover. At minimum, it should not treat "not a union" as license to skip the relationship check entirely — the current code reads as a fallthrough bug, not a deliberate design choice, since it makes the `len(members) > 1` branch's relationship check pointless for the (far more common) non-union case.

### LINKED ISSUE #221: redundant-type-conversion (TR6): silently degraded ty analysis (e.g. workspace-root mismatch) makes the diagnostics-diff always look "safe"
## Summary

`decide_candidates()` (`src/pre_commit_hooks/ast_checks/redundant_type_conversion/analysis.py`) decides a conversion is redundant by comparing `ty`'s diagnostics for the file before and after removing the wrap: if `after - baseline` is empty, it reports the conversion as redundant. This inference is only sound if `ty` is actually able to fully analyze the file. I found a case where it silently isn't, and the diff then always comes back empty regardless of the real type relationship.

## Repro

Two files, identical content, byte-for-byte:

```py
# frozenset->set conversion feeding .update(), then tuple(...) feeding random.choice(...)
# (full repro available on request; matches the shape from issue #220)
```

- Analyzed via a `ty` session/daemon **rooted at the file's own directory**: `ty` correctly reports `unresolved-attribute` when the `set(...)` wrap is removed (frozenset has no `.update`), and `invalid-argument-type` when the `tuple(...)` wrap is removed (`random.choice` requires `SupportsLenAndGetItem`, not `set`). Both conversions are correctly judged *not* redundant.
- Analyzed via a `ty` daemon **rooted at an unrelated project directory** (the file lives outside that root): `session.open_or_update()` returns **zero diagnostics** for the file in both the "before" and "after" states — `ty` isn't actually checking it. The diagnostics-diff is then empty by construction, and both conversions get reported as "redundant," with the same misleading "`ty` sees no difference with or without this wrap" message that's supposed to mean "verified safe."

I reproduced this directly against the daemon/session objects (`TySession.open_or_update`, `daemon.connect`), not just through the CLI, so this isn't a fluke of test setup — it's `ty`'s own behavior when a file falls outside its discovered project scope.

## Why this is a real risk, not just a synthetic edge case

`get_session()` → `_acquire_session()` → `daemon.connect(Path.cwd())` roots the persistent daemon at the **process's cwd** (well, the nearest ancestor `.git`), not at the target file's own directory or its own nearest `ty`/`pyproject.toml` config. Any setup where the invoking cwd's discovered `ty` root doesn't fully cover the file being checked — a monorepo with nested `pyproject.toml`/`ty.toml` scoping, a file outside the discovered project's `include`, pre-commit/prek being invoked from a different directory than expected, stale daemon reuse after a workspace layout change — reproduces this silently. There's no cross-check that `ty` actually analyzed the file at all before trusting an empty diff as "no difference."

This looks related to the "isolate ty configuration discovery" work in 4946146, but that commit specifically scopes the *self-test's* ty config so it isn't polluted by the consumer's project — it doesn't address the consumer-facing analysis session picking up a `ty` root that under-covers the actual target file.

## Suggested fix

Before trusting `after - baseline == empty` as proof of redundancy, verify `ty` actually produced *some* meaningful signal for the file (e.g. it's not silently empty for both states when the file has real content ty should be able to check), or root the analysis session using the target file's own nearest `ty`/`pyproject.toml` config rather than (or in addition to) `Path.cwd()`. At minimum, this should be documented as a known-unsound condition so callers can recognize it.

### LINKED ISSUE #222: redundant-type-conversion (TR6): no f-string-splice guard, so removing a wrap can change what actually gets interpolated
## Summary

`redundant-type-conversion` decides redundancy purely from `ty`'s type diagnostics (see #221 for a case where those diagnostics themselves go missing). But even when `ty`'s type analysis is fully correct, it fundamentally can't see this class of bug: two types can be interchangeable from a type-checker's point of view while producing different `repr()`/`str()` output when interpolated into an f-string. `redundant_assignment` already has a dedicated guard for exactly this (`fstring_splice_is_safe` / `is_safe_to_splice_into_fstring` in `redundant_assignment/semantic.py`); `redundant_type_conversion` has no equivalent.

## Repro

```py
def check_origin_airports_have_flights(location: SnapshotLocation, iatas: frozenset[IataAirportCode]) -> bool:
    iatas_for_sql = tuple(iatas)
    sql = f"""
            SELECT EXISTS (
                SELECT 1
                FROM enriched_flights
                WHERE origin_iata IN {iatas_for_sql} OR destination_iata IN {iatas_for_sql}
            ) AS contains_any;
            """
```

```
TR6: Redundant `tuple(...)` conversion: `ty` sees no difference with or without this wrap here, though the argument's own type is `frozenset[IataAirportCode]`, not `tuple` -- verify before removing.
```

Removing `tuple(...)` here is not type-redundant: DuckDB's `IN (...)` clause needs the SQL literal that `tuple.__repr__` produces (`('A', 'B')`); `frozenset.__repr__` produces `frozenset({'A', 'B'})`, which is not valid SQL. `ty` correctly reports no type error either way — both a `frozenset` and a `tuple` are perfectly valid to embed in an f-string — so this passes the diagnostics-diff cleanly while still changing runtime behavior.

## Suggested fix

Before reporting a candidate as redundant, check whether the wrapped call's result (or, transitively, the variable it's assigned to) is later spliced into an f-string / `str.format()` / `%`-formatting, similar to how `redundant_assignment`'s `fstring_splice_is_safe` already reasons about f-string safety for its own inlining. If so, skip the report (or require the exact-match, non-aggressive path only) since container `repr()` differences are invisible to `ty`'s diagnostics.

### PR DESCRIPTION
**Pull request by alessio-locatelli**
Closes https://github.com/alessio-locatelli/ruff-extra-rules/issues/220

Closes https://github.com/alessio-locatelli/ruff-extra-rules/issues/221

Closes https://github.com/alessio-locatelli/ruff-extra-rules/issues/222

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved redundant type-conversion detection in aggressive mode by requiring a plausible relationship between the value and constructor types.
  - Skips files outside the project root with a warning when diagnostics are unavailable.
  - Avoids reporting conversions used in f-strings, `str.format()`, or `%` formatting unless the type match is exact.

- **Documentation**
  - Updated rule documentation to describe project-root handling and string-formatting safeguards.

### CODE REVIEW

File: src/pre_commit_hooks/ast_checks/redundant_type_conversion/confidence.py [hunk 1]
```diff
@@ -80,6 +80,61 @@ def is_purepath_hover(hover_text: str) -> bool:
...
+    "int": frozenset({"int", "bool"}),
+    "float": frozenset({"float", "int"}),
+    "bool": frozenset({"bool"}),
+}
+
+
+def _loose_match(hover_text: str, constructor: str) -> bool:
+    head = hover_text.split(" & ", 1)[0].split("[", 1)[0]
+    if constructor in _ITERABLE_CONSTRUCTORS:
+        return head in _ITERABLE_PROTOCOL_HEADS
```
[the remark this answers is not in the document]
Comment (alessio-locatelli) [reply src/pre_commit_hooks/ast_checks/redundant_type_conversion/confidence.py:134]: Committed as `c955148`. Two of the three inline findings were valid bugs and are now fixed: `bytearray(str)` is no longer accepted as a loose aggressive-mode match (it raises `TypeError` at runtime without an explicit encoding), and `.format()` calls on non-string receivers no longer suppress valid reports. The third suggestion (scope/order-aware alias tracking) was skipped — it asks this guard to behave differently from the file's existing, deliberately whole-file-scoped heuristics (e.g. `shadowed`), and since this guard only ever suppresses (never wrongly promotes) a report, the imprecision is safe by design. Full suite, coverage, mypy, and the aggressive self-check all pass clean.

**Comment by alessio-locatelli** [comment]
@coderabbitai review full
